### ISSUE TITLE
Git changes expand to "actual" paths when opened through a `subst` drive location

### ISSUE THREAD

**Issue by alvinhochun** [microsoft/vscode#100274]
Issue Type: <b>Bug</b>

1. Create a subst drive (`mkdir C:\subst_dir && subst Z: C:\subst_dir`)
2. Create a git repo in there (`mkdir Z:\repo && cd /d Z:\repo && git init && echo Hello > test.txt && git add test.txt && git commit -m first`)
3. Open the repo in VSCode (`code repo`)
4. Make some changes to the file and save.

Bug effects (non-exhaustive):

- The file shown on the explorer sidebar does not show as modified. (Expected results: there should be an "M" annotation and the filename should be in orange, the tooltip should also say "Modified" at the end.)
- The source control sidebar lists changes with the full "actual" path `C:\subst_dir\repo\test.txt` (Expected results: it should just show `test.txt` without the full path, and the tooltip will show the full path as `Z:\repo\test.txt`)
- When the file is opened, pressing the "Open Changes" button on the toolbar (top-right) does not navigate to the git changes view.
- When the git changes view is opened from the source control sidebar, pressing the "Open File" button on the toolbar opens the file on the "actual" path instead of the file in the workspace. If the workspace file was already opened, there will now be two tabs for the same file but with a different path.

This starts happening only after I recently updated Git for Windows to `git version 2.27.0.windows.1`. Unfortunately, I don't remember which version I updated it from (it was probably a few months old). I do have another  system with a rather old version (`git version 2.16.2.windows.1`) that does not show this behaviour even on lastest stable VSCode.

VS Code version: Code 1.46.0 (a5d1cc28bb5da32ec67e86cc50f84c67cc690321, 2020-06-10T09:03:20.462Z)
OS version: Windows_NT x64 10.0.18363

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz (8 x 1992)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off_ok<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|23.86GB (14.54GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>Extensions: none

**Comment by alvinhochun** [comment]
Update: I tested some of the older releases of Git for Windows and confirmed that it worked as expected on 2.24.1(2) and fails starting from 2.25.0.

**Comment by gjsjohnmurray** [comment]
cc @joaomoreno

**Comment by joaomoreno** [comment]
@alvinhochun What about 2.27.0?

**Comment by alvinhochun** [comment]
It seems that the apparent cause is `git rev-parse --show-toplevel` expanding through the `subst` drive location starting from git version 2.25.0.

**Comment by alvinhochun** [comment]
Looks like this issue is similar to https://github.com/microsoft/vscode/issues/91548, which was fixed by https://github.com/microsoft/vscode/commit/68e7b61f918b8d6013d0320d200e55d91530f03d using a block of seemingly hacky code. The fix does not handle other types of path canonicalization though.

@eamodio Since you fixed the other issue, do you have any comments on this?

**Comment by alvinhochun** [comment]
Should `git rev-parse --show-cdup` be used instead since it returns a relative path? But it might not work if a sub-directory in a git repo (as opposed to one of its parent directories) is linked to a drive letter.

**Comment by stangelandm1** [comment]
workaround to return the top path via --show-cdup and realpath:

git rev-parse --show-cdup | xargs realpath

**Comment by AustinWise** [comment]
You can also reproduce this problem on using symbolic links.

```bash
C:\>ver

Microsoft Windows [Version 10.0.19042.685]

C:\>wsl -d Ubuntu-18.04
/mnt/c$ cd
~$ uname -r
4.19.128-microsoft-standard
~$ git --version
git version 2.17.1
~$ mkdir realgit
~$ cd realgit/
~/realgit$ git init
Initialized empty Git repository in /home/austin/realgit/.git/
~/realgit$ echo test > test.txt
~/realgit$ git add test.txt
~/realgit$ git commit -m inital
[master (root-commit) 9ec21ef] inital
 1 file changed, 1 insertion(+)
 create mode 100644 test.txt
~/realgit$ cd
~$ ln -s realgit fakegit
~$ cd fakegit
~/fakegit$ git rev-parse --show-toplevel
/home/austin/realgit
```

If you open `~/realgit` with VS Code, it works correctly. If you open `~/fakegit` with VS Code, it exhibits the same problems described in this bug.

I also tried a 2.27 version of Git inside docker and `git rev-parse --show-toplevel` behaves the same.

FYI, I think the relevant Git for Windows issue that prompted the behavior change on the Windows side was git-for-windows/git#2481.

**Comment by roccokreutz** [comment]
I as well would like this to be fixed.
Because of this gitlens and other stuff does not work properly (e.g. no line history - author who changed line - does work)

**Comment by johannes-luebke** [comment]
Is there any update on if this will be fixed at some point?
Are there any workarounds?

The fact we're unable to use **subst** reliably creates some really big issues for us.
We use subst to ensure the same path for every user on every machine regardless of their local drives.

I see some other tools had the same problem, but were able to fix it.
https://github.com/jpmorganchase/modular/pull/787

**Comment by avatards** [comment]
A similar unwanted behavior is with VS Code itself (no git) opening a separate, additional file editor on the expanded path with the "Go to definition" command for files on a subst drive even when the workspace is set to the subst drive folder and the subst source file with the definition is already open. The files are then treated as different ones by the workspace, e.g. breakpoints which are set/removed in the subst file are not propagated to the expanded-path version and vice versa.

Not sure I need to start another issue for this since it seems related.

**Comment by lszomoru** [comment]
We have done some improvements to the git extension when it comes to file paths so subst drives should work as expected. Could any of you please download VS Code Insiders 1.99, give it a try and let me know if you still see any issues. Thanks!

**Comment by avikpram** [comment]
Any update on this? Can we expect a fix (or a workaround) in the May release?

**Comment by lszomoru** [comment]
@avikpram, what version of VS Code are you using? What issues are you running into? Thanks!

**Comment by avikpram** [comment]
@lszomoru Version I am using:
```
Version: 1.101.0 (user setup)
Commit: dfaf44141ea9deb3b4096f7cd6d24e00c147a4b1
Date: 2025-06-11T15:00:50.123Z
Electron: 35.5.1
ElectronBuildId: 11727614
Chromium: 134.0.6998.205
Node.js: 22.15.1
V8: 13.4.114.21-electron.0
OS: Windows_NT x64 10.0.22631
```

Create substitute drive `X` pointing to `C:\work\git\scripts\`
```
subst X: C:\work\git\scripts\
```

Issues noticed:

- No file shown on the explorer sidebar! `OPEN EDITORS`, `TIMELINE`, and `OUTLINE` sections show the correct content.
- The source control sidebar shows changes with the path `C:\work\git\scripts\file.txt` instead of `X:\file.txt`
- When the file-diff view is opened from the source control sidebar, breadcrumbs show the `C:\work\git\scripts` path instead of the substituted path.
- In the source control sidebar, pressing "Open File" opens the file on the actual path instead of the substituted path.

Additionally, inline git blame preview on hover (provided by Gitlens) and similar features do not work.

**Comment by mithom** [comment]
any update on this? around the same time as the last comment i had created an issue, but i forgot to follow up on it, so it got auto closed.
i really would like to have this working as any click-through action from the LSP opens the subst drive version of the file, which don't have any of the git integrations working as described by avik above.
